### PR TITLE
Make error when internal tenant ID is empty more recognisable

### PR DESCRIPTION
**Pull request by kjaksik**
**Description**

Changes proposed in this pull request:

- Add `TenantNotFound` and `TenantIsRequired` errors

**Related issue(s)**
#1287

### CODE REVIEW

File: components/director/pkg/apperrors/errors.go [hunk 1]
```diff
@@ -265,3 +265,55 @@ func IsInsufficientScopes(err error) bool {
...
+
+type emptyTenantError struct {
+}
+
+func NewEmptyTenantError() *emptyTenantError {
+	return &emptyTenantError{}
+}
+
+func (e *emptyTenantError) Error() string {
+	return "internal tenantID is empty"
```
Comment (PK85) [inline components/director/pkg/apperrors/errors.go:307]: Tenant not found.

File: components/director/internal/authenticator/middleware.go [hunk 2]
```diff
@@ -120,8 +125,9 @@ func (a *Authenticator) getBearerToken(r *http.Request) (string, error) {
 
 func (a *Authenticator) contextWithClaims(ctx context.Context, claims Claims) context.Context {
 	ctxWithTenant := tenant.SaveToContext(ctx, claims.Tenant)
```
Comment (crabtree) [inline components/director/internal/authenticator/middleware.go:127]: What about refactoring this in a way to handle tenant IDs in a struct and storing this struct in the context?

File: components/director/internal/domain/oauth20/service.go [hunk 3]
```diff
@@ -51,7 +53,9 @@ func NewService(scopeCfgProvider ScopeCfgProvider, uidService UIDService, cfg Co
 func (s *service) CreateClientCredentials(ctx context.Context, objectType model.SystemAuthReferenceObjectType) (*model.OAuthCredentialDataInput, error) {
 	scopes, err := s.getClientCredentialScopes(objectType)
 	if err != nil {
-		return nil, err
+		if !(apperrors.IsEmptyTenant(err) && objectType == model.IntegrationSystemReference) {
```
Comment (crabtree) [inline components/director/internal/domain/oauth20/service.go:56]: Can you refactor this if-statement to be more descriptive? Maybe just extract this part to some simple function that will return a bool and named it to reveal the actual intention. I also see similar patter in multiple lines so please align the other places.

File: components/director/internal/authenticator/middleware.go [hunk 4]
```diff
@@ -64,6 +64,11 @@ func (a *Authenticator) Handler() func(next http.Handler) http.Handler {
 				return
 			}
 
+			if claims.Tenant == "" && claims.ExternalTenant != "" {
+				a.writeError(w, fmt.Sprintf("No tenant found for external tenant: %s", claims.ExternalTenant), http.StatusBadRequest)
```
[the remark this answers is not in the document]
Comment (tgorgol) [reply components/director/internal/authenticator/middleware.go:68]: Also, please log an error.

File: components/director/internal/authenticator/middleware.go [hunk 5]
```diff
@@ -64,6 +64,13 @@ func (a *Authenticator) Handler() func(next http.Handler) http.Handler {
 				return
 			}
 
+			if claims.Tenant == "" && claims.ExternalTenant != "" {
+				msg := fmt.Sprintf("No tenant found : %s", claims.ExternalTenant)
```
Comment (tgorgol) [inline components/director/internal/authenticator/middleware.go:68]:
```suggestion
				msg := fmt.Sprintf("No tenant found: %s", claims.ExternalTenant)
```

File: components/director/internal/model/system_auth.go [hunk 6]
```diff
@@ -50,3 +53,10 @@ const (
 	ApplicationReference       SystemAuthReferenceObjectType = "Application"
 	IntegrationSystemReference SystemAuthReferenceObjectType = "Integration System"
 )
+
+func IsIntegrationSystemNoTenantFlow(err error, objectType SystemAuthReferenceObjectType) bool {
+	if apperrors.IsEmptyTenant(err) && objectType == IntegrationSystemReference {
+		return true
+	}
+	return false
+}
```
Comment (tgorgol) [inline components/director/internal/model/system_auth.go:62]:
```suggestion
func IsIntegrationSystemNoTenantFlow(err error, objectType SystemAuthReferenceObjectType) bool {
	return apperrors.IsEmptyTenant(err) && objectType == IntegrationSystemReference
}
```

File: components/director/internal/authenticator/middleware.go (hunk 5 above)
Comment (crabtree) [inline components/director/internal/authenticator/middleware.go:68]:
```suggestion
				msg := fmt.Sprintf("Tenant not found: %s", claims.ExternalTenant)
```

File: tests/director/gateway-integration/tenants_test.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/kyma-incubator/compass/tests/director/pkg/idtokenprovider"
+	"github.com/kyma-incubator/compass/tests/director/pkg/ptr"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	wrongTenant        = "b1e46bd5-18ba-4a02-b96d-631a9e803504"
+	emptyTenant        = ""
+	noTenantMessage    = "tenant is required"
+	wrongTenantMessage = "No tenant found :"
```
Comment (crabtree) [inline tests/director/gateway-integration/tenants_test.go:18]:
```suggestion
	wrongTenantMessage = "Tenant not found:"
```

File: docs/director/03-error-handling.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+Director API returns custom errors if anything goes wrong while handling the request. The errors can be divided into
+two categories: regular HTTP errors and GraphQL errors.
+
+## Regular Errors
+
+These errors are returned when something is wrong with the request eg. missing headers.
+
+- Tenant not found
+    - Tenant does not occur in the database
```
Comment (klaudiagrz) [inline docs/director/03-error-handling.md:11]:
```suggestion
This error occurs either when there is no tenant in the database, or when the passed tenant does not match the tenant in the [System Auth](https://github.com/kyma-incubator/compass/blob/732486482e1d71384be4d705b1ed260365f74c1c/docs/compass/03-01-security.md#system_auths-table). 
```

File: docs/director/03-error-handling.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    - In case of Application/Runtime flow the passed tenant does not match the tenant in System Auth
```
Comment (klaudiagrz) [inline docs/director/03-error-handling.md:12]:
```suggestion

```

File: docs/director/03-error-handling.md [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+
+## GraphQL Errors
+
+These errors are returned eg. when there was a problem with connecting to the database, or in case an inter server error occured.GraphQL.
+
+- Tenant is required
```
Comment (klaudiagrz) [inline docs/director/03-error-handling.md:19]:
```suggestion
**Tenant is required**
```

File: docs/director/03-error-handling.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+These errors are returned when something is wrong with the request eg. missing headers.
+
+- Tenant not found
+    - Tenant does not occur in the database
+    - In case of Application/Runtime flow the passed tenant does not match the tenant in System Auth
+
+
+## GraphQL Errors
+
+These errors are returned eg. when there was a problem with connecting to the database, or in case an inter server error occured.GraphQL.
```
Comment (klaudiagrz) [inline docs/director/03-error-handling.md:17]:
```suggestion
GraphQL errors are related to the connection. They may concern issues with connecting to the database, or inter-server issues. These are the GraphQL errors that may occur when calling the Director API:
```

File: docs/director/03-error-handling.md [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+- Tenant is required
+    - No tenant was passed but it's required by the operation
```
Comment (klaudiagrz) [inline docs/director/03-error-handling.md:20]:
```suggestion
This error occurs when a tenant, that is required by the operation, wasn't passed.
```

File: docs/director/03-error-handling.md [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+Director API returns custom errors if anything goes wrong while handling the request. The errors can be divided into
+two categories: 
+- Regular HTTP errors 
+- GraphQL errors
+
+## Regular errors
+
+Regular errors are related to the request itself. They may concern issues such as missing headers. These are the regular errors that may occur when calling the Director API:
+
+**Tenant not found**
```
Comment (klaudiagrz) [inline docs/director/03-error-handling.md:12]:
```suggestion
**Tenant not found**

```

File: docs/director/03-error-handling.md [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+This error occurs either when there is no such tenant in the database, or when the passed tenant does not match the tenant in the [System Auth](https://github.com/kyma-incubator/compass/blob/732486482e1d71384be4d705b1ed260365f74c1c/docs/compass/03-01-security.md#system_auths-table). 
+
+
+## GraphQL errors
+
+GraphQL errors are related to the request execution. They may concern issues with connecting to the database, or inter-server issues. These are the GraphQL errors that may occur when calling the Director API:
+
+**Tenant is required**
```
Comment (klaudiagrz) [inline docs/director/03-error-handling.md:20]:
```suggestion
**Tenant is required**

```

File: docs/director/03-error-handling.md [hunk 15]
```diff
@@ -0,0 +1,21 @@
+# Error handling
+
+Director API returns custom errors if anything goes wrong while handling the request. The errors can be divided into
```
Comment (klaudiagrz) [inline docs/director/03-error-handling.md:3]:
```suggestion
Director API returns custom errors when there are problems with handling the request. The errors can be divided into
```

**Pull request merged**